crates/matrix-sdk/src/room/timeline/event_handler.rs

```diff
@@ -0,0 +1,460 @@
+    fn from(event: AnySyncTimelineEvent) -> Self {
+        match event {
+            AnySyncTimelineEvent::MessageLike(AnySyncMessageLikeEvent::RoomRedaction(
+                SyncRoomRedactionEvent::Original(OriginalSyncRoomRedactionEvent {
+                    redacts,
+                    content,
+                    ..
+                }),
+            )) => Self::Redaction { redacts, content },
+            AnySyncTimelineEvent::MessageLike(ev) => match ev.original_content() {
+                Some(content) => Self::Message { content },
+                None => Self::RedactedMessage,
+            },
+            AnySyncTimelineEvent::State(ev) => match ev.original_content() {
+                Some(_content) => Self::State { _content },
+                None => Self::RedactedState,
+            },
+        }
+    }
+}
+
+#[derive(Clone)]
+enum EventKind {
+    Message { content: AnyMessageLikeEventContent },
+    RedactedMessage,
+    Redaction { redacts: OwnedEventId, content: RoomRedactionEventContent },
+    // FIXME: Split further for state keys of different type
+    State { _content: AnyStateEventContent },
+    RedactedState, // AnyRedactedStateEventContent
+}
+
+enum EventPosition {
+    Start,
+    End,
+}
+
+struct EventHandler<'a> {
+    meta: EventMetadata,
+    flow: Flow,
+    timeline: &'a TimelineInner,
+    event_added: bool,
+}
+
+impl<'a> EventHandler<'a> {
+    fn new(meta: EventMetadata, flow: Flow, timeline: &'a TimelineInner) -> Self {
+        Self { meta, flow, timeline, event_added: false }
+    }
+
+    fn handle_event(mut self, event_kind: EventKind) {
+        match event_kind {
+            EventKind::Message { content } => match content {
+                AnyMessageLikeEventContent::Reaction(c) => self.handle_reaction(c),
+                AnyMessageLikeEventContent::RoomMessage(c) => self.handle_room_message(c),
+                // TODO
+                _ => {}
+            },
+            EventKind::RedactedMessage => {
+                self.add(NewEventTimelineItem::redacted_message());
+            }
+            EventKind::Redaction { redacts, content } => self.handle_redaction(redacts, content),
+            // TODO: State events
+            _ => {}
+        }
+
+        if !self.event_added {
+            // TODO: Add event as raw
+        }
+    }
+
+    fn handle_room_message(&mut self, content: RoomMessageEventContent) {
+        match content.relates_to {
+            Some(Relation::Replacement(re)) => {
+                self.handle_room_message_edit(re);
+            }
+            _ => {
+                self.add(NewEventTimelineItem::message(content, self.meta.relations.clone()));
+            }
+        }
+    }
+
+    fn handle_room_message_edit(&mut self, replacement: Replacement) {
+        let event_id = &replacement.event_id;
+
+        self.maybe_update_timeline_item(event_id, "edit", |item| {
+            if self.meta.sender != item.sender() {
+                info!(
```

Comment:
All the log lines seem to be missing the room id, this will be quite annoying to debug if we don't know the room the log line applies to.

Reply:
I've added it for local echo and back-pagination, for live events we should just add a useful span for all event handlers, not just the timeline ones.